docs/en/integrations/language-clients/rust.md

```diff
@@ -0,0 +1,517 @@
+
+```rust
+let numbers = client
+    .query("SELECT number FROM system.numbers LIMIT 1")
+    .with_option("query_id", "some-query-id")
+    .fetch_all::<u64>()
+    .await?;
+```
+
+Besides `query`, it works similarly with `insert` and `inserter` methods.
+
+:::danger
+If you set `query_id` manually, make sure that it is unique. UUIDs are a good choice for this.
+:::
+
+See also: [query_id example](https://github.com/ClickHouse/clickhouse-rs/blob/main/examples/query_id.rs) in the client repo.
+
+### Session ID
+
+Similarly to `query_id`, you can set the `session_id` to execute the statements in the same session. `session_id` can be set either globally on the client level, or per `query`, `insert`, or `inserter` call.
+
+```rust
+let client = Client::default()
+    .with_url("http://localhost:8123")
+    .with_option("session_id", "my-session");
+```
+
+:::danger
+With clustered deployments, due to lack of "sticky sessions", you need to be connected to a _particular cluster node_ in order to properly utilize this feature, cause, for example, a round-robin load-balancer will not guarantee that the consequent requests will be processed by the same ClickHouse node.
+:::
+
+See also: [session_id example](https://github.com/ClickHouse/clickhouse-rs/blob/main/examples/session_id.rs) in the client repo.
+
+### Custom HTTP headers
+
+If you are using proxy authentication or need to pass custom headers, you can do it like this:
+
+```rust
+let client = Client::default()
+    .with_url("http://localhost:8123")
+    .with_header("X-My-Header", "hello");
+```
+
+See also: [custom HTTP headers example](https://github.com/ClickHouse/clickhouse-rs/blob/main/examples/custom_http_headers.rs) in the client repo.
+
+### Custom HTTP client
+
+This could be useful for tweaking the underlying HTTP connection pool settings.
+
+```rust
+use hyper_util::client::legacy::connect::HttpConnector;
+use hyper_util::client::legacy::Client as HyperClient;
+use hyper_util::rt::TokioExecutor;
+
+let connector = HttpConnector::new(); // or HttpsConnectorBuilder
+let hyper_client = HyperClient::builder(TokioExecutor::new())
+    // For how long keep a particular idle socket alive on the client side (in milliseconds).
+    // It is supposed to be a fair bit less that the ClickHouse server KeepAlive timeout,
+    // which was by default 3 seconds for pre-23.11 versions, and 10 seconds after that.
+    .pool_idle_timeout(Duration::from_millis(2_500))
+    // Sets the maximum idle Keep-Alive connections allowed in the pool.
+    .pool_max_idle_per_host(4)
+    .build(connector);
+
+let client = Client::with_http_client(hyper_client).with_url("http://localhost:8123");
+```
+
+:::warning
+This example relies on the legacy Hyper API and is a subject to change in the future.
+:::
+
+See also: [custom HTTP client example](https://github.com/ClickHouse/clickhouse-rs/blob/main/examples/custom_http_client.rs) in the client repo.
+
+## Data Types
```

Comment:
should we add a table with supported type, as you did in https://clickhouse.com/docs/en/integrations/javascript#supported-clickhouse-data-types ?

Reply:
Yes, that is ideally a part of the rework in the follow-ups.